src/cswinrt/code_writers.h

```diff
@@ -6340,8 +6419,117 @@ type.TypeNamespace(),
 type.TypeName(),
 bind<write_type_name>(type, typedef_name_type::CCW, true)
 );
-},
-types
-));
+    },
+    types
+        ));
     }
-}
\ No newline at end of file
+
+    void write_event_source_generic_args(writer& w, cswinrt::type_semantics eventTypeSemantics)
+    {
+        if (!std::holds_alternative<generic_type_instance>(eventTypeSemantics))
+        {
+            return;
+        }
+        for_typedef(w, eventTypeSemantics, [&](TypeDef const& eventType)
+            {
+                std::vector<std::string> genericArgs;
+                int i = 0;
+                for (auto&& x: std::get<generic_type_instance>(eventTypeSemantics).generic_args)
+                {
+                    auto semantics = w.get_generic_arg_scope(i).first;
+                    if (std::holds_alternative<generic_type_param>(semantics))
+                    {
+                        genericArgs.push_back(w.write_temp("%", bind<write_generic_type_name>(i)));
+                    }
+                    i++;
+                }                
+                if (genericArgs.size() == 0)
+                {
+                    return;
+                }
+                w.write("<%>", bind_list([](writer& w, auto&& value)
+                    {
+                        w.write(value);
+                    }, ", "sv, genericArgs));
+            });
+    }
+
+    void write_event_source_subclass(writer& w, cswinrt::type_semantics eventTypeSemantics)
+    {
+        for_typedef(w, eventTypeSemantics, [&](TypeDef const& eventType)
+            {
+                if (eventType.TypeNamespace() == "System" && eventType.TypeName() == "EventHandler`1")
+                {
+                    return;
+                }
+                auto eventTypeCode = w.write_temp("%", bind<write_type_name>(eventType, typedef_name_type::Projected, false));
+                auto invokeMethodSig = get_event_invoke_method(eventType);
+                w.write(R"(
+    internal unsafe class %% : EventSource<%>
+    {
+        private % handler;
+
+        internal %(IObjectReference obj,
+            delegate* unmanaged[Stdcall]<System.IntPtr, System.IntPtr, out WinRT.EventRegistrationToken, int> addHandler,
+            delegate* unmanaged[Stdcall]<System.IntPtr, WinRT.EventRegistrationToken, int> removeHandler) : base(obj, addHandler, removeHandler)
+        {
+        }
+
+        override protected System.Delegate EventInvoke
+        {
+            get
+            {
+                if (handler == null)
```

Comment:
I wonder if we should just set handler in the constructor to avoid the null check.  But we can also leave it as is.

Reply:
We just save having to create a handler object when not needed. This code is basically doing exactly what the reflection code was doing.